### PR TITLE
etc/ file installation fix, master branch (2018.02.13.)

### PR DESCRIPTION
**Pull request by krasznaa**
This almost broke me...

I've been trying to find for days why ROOT would not build correctly in one particular (fairly complicated) setup for me. I just couldn't figure out why I wouldn't get the files meant for the `etc/` directory in my installation.

Turns out that it was because in this weird setup I put the ROOT source code into a directory inside of another CMake project, into a directory that had `/CMakeFiles/` as part of its path. And would you know, macOS still comes with a case insensitive file system, so this exclusion rule was triggering for the full contents of the `etc/` directory in this setup.

After realising this, I modified my project's setup to put the ROOT source code into a safer location. But this update should still help people avoid a similar situation later on...

P.S. I checked in a simple standalone example that this formalism should do the right thing. (Took me some tries to find "just the right" incantation.)

### CODE REVIEW

**Comment by krasznaa** [comment]
Now that I slept on it...

As long as we're updating this `install(...)` command, we probably want to update all exclusion rules to be as specific as possible. Should I update the pull request like that?

Cheers,
        Attila

File: CMakeLists.txt [hunk 1]
```diff
@@ -418,7 +418,7 @@ if(NOT CMAKE_SOURCE_DIR STREQUAL CMAKE_INSTALL_PREFIX)
                          REGEX rootd.xinetd EXCLUDE
                          REGEX proofd.xinetd EXCLUDE
                          REGEX root.mimes EXCLUDE
-                         REGEX cmake EXCLUDE
+                         REGEX "/cmake$" EXCLUDE
```
Comment (amadio) [inline CMakeLists.txt]: Does this still exclude installing the `cmake` subdirectory under `etc/cmake`? Also, these things seem to be inverted, as the files above are meant to be installed. @Axel-Naumann, any comments?

Comment (krasznaa) [reply CMakeLists.txt]: Yes, it does. This is what I was playing with a bit. If I specified for instance "/cmake/", that would exclude all files from inside of the `cmake/` directory to be installed, but would've still created an empty `cmake/` directory in the installation area.

Now... Whether the two files that are in `etc/cmake` should be installed along with the release or not, is a fair question. I always found it weird that you wouldn't ship `FindROOT.cmake` with the installed ROOT release. Though the other file, `RootUseFile.cmake`, **is** installed with the release through another path. So that should definitely be excluded here.

If you check, many of the referenced files are not actually part of the `etc/` source directory. I believe those files are just being put into the `etc/` directory in the build area during the build. As such, setting up an exclusion rule for them is unnecessary here. Since when you refer to directory `foo/` without an absolute path, CMake should take that as `${CMAKE_CURRENT_SOURCE_DIR}/foo/`.

But I guess this does make sense for in-source builds... So okay, it may be worth keeping these after all. 😛 

Long story short, I would probably just extend every exclusion rule with a `/` prefix and a `$` postfix...

Comment (krasznaa) [reply CMakeLists.txt]: And since the files are being generated during the build, their installation is set up "elsewhere" in the configuration. Since its very unhealthy to install the same file in two or more "different ways", the exclusion rules here are very warranted.

Comment (amadio) [reply CMakeLists.txt]: `ROOTConfig.cmake` replaces `FindROOT.cmake`. You only need the second one if the first is not provided. See https://cmake.org/cmake/help/v3.10/command/find_package.html for more information.

Comment (krasznaa) [reply CMakeLists.txt]: I'm quite familiar with CMake's package finding infrastructure. 😜

These days, when ROOT can only be built by CMake (is that the case by now?), this is a fair argument. But then I wonder why that file is still in the repository...

P.S. For software projects that can use both ROOT 5 and ROOT 6, it's usually still a good idea to have a `FindROOT.cmake` module of their own. Though those anyway have to be implemented differently from what's here. Like:

https://gitlab.com/atomki-nuclear-phys/cda/blob/master/cmake/FindROOT.cmake

Still, even for this, I did take inspiration from the `FindROOT.cmake` module of this repository. So you guys may still want to advertise it a bit more...

Comment (amadio) [reply CMakeLists.txt]: Sorry, I understand your question now. :-)

However, why are dependent projects using their own (or our own) `FindROOT.cmake` instead of relying on the `ROOTConfig.cmake`? Are there missing features? Maybe we need to file them as bugs and fix, and when `ROOTConfig.cmake` satisfies dependent projects, we can get rid of `FindROOT.cmake` everywhere. What do you think?

Comment (amadio) [reply CMakeLists.txt]: BTW, since we already have `/http` as pattern for exclusion, would `/cmake` rather than `/cmake$` also work? If yes, please update the commit and I will merge. Thanks!

Comment (krasznaa) [reply CMakeLists.txt]: 😕 If you want to build against a ROOT (5) version that was built using `configure`, you have to use a `FindROOT.cmake` module. The example I gave should demonstrate that nicely. (As it is meant to build against Ubuntu's own ROOT version as well for instance. Which does not come with its own CMake configuration.)

No, `/cmake` would not work. That would still kill the installation of every file from the `etc/` directory if the ROOT source was sitting in any of these sort of directories:
  - `/somewhere/cmake-builds/root`;
  - `/somewhere/myProject/build/CMakeFiles/root`.

In fact I would also update the `/http` pattern, as that would also be destructive for a source directory like `/somwehre/http-downloads/root`.

As I wrote, I would prefer to add the `/` prefix and `$` postfix to **all** regular expressions.

Comment (amadio) [reply CMakeLists.txt]: I see, then we should be providing `ROOTConfig.cmake` even if built with the old `configure` script before we can get rid of `FindROOT.cmake`...

You are right about the `$` at the end, let's update both expressions as you sugest (but not the ones that are filenames). I am still surprised that the patterns consider the full path and not just the relative path within `${CMAKE_BINARY_DIR}`. Thanks for all the explanations.

Comment (krasznaa) [reply CMakeLists.txt]: See: https://cmake.org/cmake/help/v3.8/command/install.html#installing-directories

>A PATTERN will match only complete file names: the portion of the full path matching the pattern must occur at the end of the file name and be preceded by a slash. A REGEX will match any portion of the full path but it may use / and $ to simulate the PATTERN behavior.

I agree that for the file names a false-match is pretty unlikely, but this formalism should still be a bit safer than what we had so far.

Comment (krasznaa) [reply CMakeLists.txt]: And if we wanted to be **absolutely** safe, we should even escape the periods in the file names. Since those currently mean "any character" in this installation command.

Or, we could just stop using `REGEX` completely. Since none of these exclusions actually need a regular expression anyway...

But I'm willing to leave that to you guys, these updates will be "enough" for my own use case.

**Comment by amadio** [comment]
I don't think that the filenames will cause similar problems, but there is no harm, so I will just merge this after it goes through Jenkins. Thanks!

**Comment by amadio** [comment]
@phsft-bot build

File: CMakeLists.txt (hunk 1 above)
Comment (amadio) [reply CMakeLists.txt]: Ah, you are totally right, we should be using `PATTERN` rather than `REGEX`.

**Comment by amadio** [comment]
@krasznaa As per our [discussion](https://github.com/root-project/root/pull/1625/commits/4ef838c3dc1629f8e7a688679854fb490eaffa09), could you please convert your patch to use `PATTERN` instead of `REGEX`, then we don't need the `$` and `/` when listing the exclusions.

**Comment by krasznaa** [comment]
Sorry for the delay. I believe this should be behaving exactly the same as the previous version. But let's see what the full suite of tests say.

**Comment by amadio** [comment]
Thanks!

**Comment by amadio** [comment]
@phsft-bot build
